framework/src/packets/mod.rs

```diff
@@ -154,6 +173,92 @@ pub trait Packet {
     }
 }
 
+/// Immutable smart pointer to a packet
+///
+/// A smart pointer that prevents the packet from being modified. The main
+/// use is allow safe lookahead of packet payload while retaining ownership
+/// of the original packet. The lifetime of the smart pointer is constrained
+/// by the original packet.
+pub struct Immutable<'a, T: Packet + 'a> {
+    value: T,
+    phantom: PhantomData<&'a T>,
+}
+
+impl<'a, T: Packet> Immutable<'a, T> {
+    pub fn new(value: T) -> Self {
+        Immutable {
+            value,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<'a, T: Packet> Deref for Immutable<'a, T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.value
+    }
+}
+
+/// Conditional reference counted smart pointer
+///
+/// The content of the pointer will be deep-copied the first time `clone` is
+/// invoked. Subsequent calls to `clone` will clone a `std::rc::Rc` pointer,
+/// and not perform deep copy any more.
+#[doc(hidden)]
+#[derive(Debug)]
+pub(crate) enum CondRc<T: Packet> {
+    Raw(T),
+    Counted(Rc<T>),
+}
+
+impl<T: Packet> CondRc<T> {
+    pub fn new(value: T) -> Self {
+        CondRc::Raw(value)
+    }
+
+    pub fn into_owned(self) -> T {
```

Comment:
@drunkirishcoder into_owned b/c `to_*` functions usually take in references (via clippy docs), where this is taking in our non-referenced enum and unwrapping it.

Reply:
yep `into_*` is the right convention here, good call.